src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetProjectManagerService.cs

```diff
@@ -99,27 +100,39 @@ public async ValueTask<IProjectContextInfo> GetProjectAsync(string projectId, Ca
             return await ProjectContextInfo.CreateAsync(project, cancellationToken);
         }
 
-        public async ValueTask<IReadOnlyCollection<IPackageReferenceContextInfo>> GetInstalledPackagesAsync(
+        /// <summary>
+        /// Returns a dictionary of projectId to installed PackageReferences.
+        /// </summary>
+        /// <param name="projectIds"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns>Dictionary of projectId to installed PackageReferences.</returns>
+        public async ValueTask<IReadOnlyDictionary<string, IReadOnlyCollection<IPackageReferenceContextInfo>>> GetInstalledPackagesAsync(
             IReadOnlyCollection<string> projectIds,
             CancellationToken cancellationToken)
         {
             Assumes.NotNullOrEmpty(projectIds);
 
             cancellationToken.ThrowIfCancellationRequested();
 
-            IReadOnlyList<NuGetProject> projects = await GetProjectsAsync(projectIds, cancellationToken);
+            ReadOnlyCollection<string> distinctProjectIds = projectIds.Distinct().ToList().AsReadOnly();
+            IReadOnlyList<NuGetProject> projects = await GetProjectsAsync(distinctProjectIds, cancellationToken);
```

Comment:
This `projects` is not passed around as parameter, so danger of mutation is very unlikely. 
So use of `IReadOnlyList` over `List` not clear, only adds tiny more overhead.

Reply:
Note that, I did not make any impacting change to this line of code in this PR.
I think the mutation could occur within this method, and this makes it clear that it cannot, and will not, mutate.